### ISSUE TITLE
B300: fuse DyT backward parameter-gradient reductions

### ISSUE THREAD

**Issue by heiheiha798** [heiheiha798/Liger-Kernel#11]
## Problem

DyT backward currently launches its autotuned Triton producer and then returns to eager PyTorch to reduce three FP32 partial-gradient buffers:

```python
if HAVE_BETA:
    db = db.sum(0).to(x.dtype)
dg = dg.sum(0).to(gamma.dtype)
da = da.sum().to(x.dtype).unsqueeze(0)
```

The partial buffers scale with the device SM count: `DG` and optional `DB` are `[NUM_SMS, N]`, while zero-initialized `DA` is `[NUM_SMS, ceil(N / 512)]`. On a B300 with 148 SMs, short and medium token counts may make the fixed sequence of reduction and low-precision cast launches a meaningful part of backward/full latency.

This is an unmeasured B300/SM103 hypothesis. The source shows a plausible launch and partial-buffer bottleneck, but profiling must establish it before any production change is retained.

Baseline at selection time: `upstream/main` at `124fb8a2a443c1118f7caeff0203aea78424eb2b`.

## Profile-first gate

In one clean B300/SM103 Slurm allocation, profile the exact baseline before making a permanent candidate commit. Use BF16 with `beta=False/True`, `M={1,16,128,2048}`, and `N={1024,3584,4096,8192}`.

Record the producer plus every subsequent reduction/cast kernel, aggregate epilogue duration, bytes where available, and contribution to directly measured backward/full time. A temporary one-launch reducer may be used for the gate, but stop and record a no-go if the eager epilogue is not present or the fixed fusion does not improve at least two representative launch-bound cases by 10% without a greater than 5% regression.

Do not tune or modify the producer to rescue the hypothesis.

## Approved implementation scope

Expected tracked files only:

- `src/liger_kernel/ops/dyt.py`
- `test/transformers/test_dyt.py`

Add one Triton reduction helper/kernel that, in one launch:

- reduces every FP32 `DG[:, col]` and optional `DB[:, col]` over the complete existing `NUM_SMS` dimension;
- reduces the complete zero-initialized `DA` storage exactly as the current `da.sum()` does, without inferring the autotune winner or reading only presumed active slots;
- stores `dalpha`, `dgamma`, and optional `dbeta` directly in their existing output shapes and parameter dtypes.

Enable it only for the actual input tensor's CUDA device when `infer_device_arch(device_id) == "blackwell_ultra"` (SM103). Resolve the device ID from `x.device`, including non-default CUDA devices. Every other CUDA architecture, ROCm, XPU, and NPU must use the current `torch.sum(...).to(...)` path unchanged.

Keep the existing `_dyt_bwd_kernel`, its 18-config autotune space, `reset_to_zero=["DA"]`, partial-buffer allocation/layout, public APIs, saved tensors, and forward path unchanged. `HAVE_BETA=False` must not read or require a valid `DB` pointer.

Do not add a second reduction kernel, autotune the epilogue, or introduce another heuristic/fallback if the single fixed design misses the performance gate.

## Correctness oracle

The primary oracle is the existing eager PyTorch DyT expression, comparing output and all gradients (`dx`, `dalpha`, `dgamma`, and optional `dbeta`) under independent non-unit random upstream gradients.

Cover:

- `beta=False/True`;
- FP32, BF16, and FP16 at existing repository tolerances;
- `M={1,2,16,128,2048}` so rows include values below the B300 SM count and a larger control;
- `N={256,1023,2048,3584,4096,8192}` for small, ragged, and production widths.

Add a focused low-level reducer oracle using artificial FP32 `DA/DG/DB` partials. Compare against the current `torch.sum(...).to(...)` sequence, including absent beta, output shapes/dtypes, and nonzero values placed in both active-looking and normally unused `DA` slots. This proves the new reducer honors the full existing storage contract rather than guessing the autotune winner.

Add dispatch tests proving exact SM103 selection uses the supplied tensor device ID and all other architectures retain the eager fallback. Run full `test/transformers/test_dyt.py` and `make checkstyle` in addition to focused coverage.

## B300 benchmark and acceptance

Final evidence must compare clean existing baseline and candidate commits with exact source attribution in the same SM103 allocation. Use identical immutable inputs/parameters/upstream gradients, warm every compiled variant, and run at least seven alternating paired rounds. Report every round plus p20/median/p80; do not report only the best run or sum independent forward/backward quantiles.

Pre-registered primary matrix (24 backward and 24 direct-full pairs):

- BF16;
- `beta=False/True`;
- `M={1,16,128}`;
- `N={1024,3584,4096,8192}`.

Controls:

- BF16 `M=2048` across the same widths and beta modes;
- FP16 and FP32 at `M={16,2048}`, `N={1024,4096,8192}`, both beta modes;
- forward latency as an unchanged-pass control;
- the repository `benchmark/scripts/benchmark_dyt.py --sweep-mode model_config --overwrite` representative model sweep as secondary evidence, without committing generated CSV changes.

Accept only if:

1. all correctness and static checks pass;
2. at least 12 of 24 primary backward paired medians improve by at least 10%;
3. at least 12 of 24 primary direct-full paired medians improve by at least 5%;
4. no primary/control shape or unchanged forward pass regresses by more than 5%; and
5. profiling confirms the expected eager epilogue launches are replaced by exactly one reduction launch.

Otherwise record a no-go without shrinking the matrix, relaxing tolerances/thresholds, or broadening scope.

## Duplicate and overlap check

No open upstream issue, PR, or remote branch implements the Triton DyT parameter-gradient epilogue fusion. Merged PR #1208 only enabled the existing producer autotune and retained `da.sum()`/`dg.sum()`/`db.sum()`.

Open PR #1321 adds an opt-in cuTile DyT backend in a separate file; this issue must not modify or benchmark that backend. Open PR #1180 is a repository-wide benchmark refactor and does not currently change the DyT script. Neither is a semantic duplicate.

## Explicit exclusions

- no producer math, `BLOCK_N`, warp/stage, autotune, grid, partial layout, or initialization changes;
- no cuTile, Ascend/NPU, XPU, ROCm, module/API, or model integration changes;
- no generic reduction abstraction or changes to other norm/activation kernels;
- no tracked benchmark script/data change, architecture expansion, additional kernel, or new fallback;
- no reuse of any prior-cycle branch or unmerged fork commit.

**Comment by heiheiha798** [comment]
## B300 profile-first result: no-go

The fixed one-launch reducer did not pass the pre-registered no-regression gate. No production source or test change was retained, and no draft PR was created.

### Source and environment

- Baseline and branch head: `124fb8a2a443c1118f7caeff0203aea78424eb2b`
- Branch: `b300/dyt-bwd-reduction-sm103`
- Slurm job: `5034` on `dev-slurm`
- GPU: NVIDIA B300 SXM6 AC, SM103, 148 SMs
- Driver: 580.126.09; CUDA runtime: 13.0
- PyTorch: 2.13.0+cu130; Triton: 3.7.1; Nsight Compute: 2025.3.1.0

The temporary ignored variant preserved `_dyt_bwd_kernel`, its autotune space, grid, and partial buffers. It replaced only the eager epilogue with one fixed Triton launch using 32 contiguous columns/program, 4 warps, a complete 148-row DG/DB reduction, and a complete DA reduction in fixed 1024-element chunks. `HAVE_BETA=False` did not read DB.

### Paired gate result

Method: BF16, `beta={False,True}`, `M={1,16,128,2048}`, `N={1024,3584,4096,8192}`, five alternating provider rounds, five warmups and 30 CUDA-event samples per provider/round. Backward and full were timed directly; both full providers used the same public `LigerDyTFunction` and contiguity wrapper. Ratios are baseline/candidate from paired round medians, so values above 1 are faster.

| Group | Shapes | Paired-median speedup range | Median across shapes | >=10% gains | >5% regressions |
|---|---:|---:|---:|---:|---:|
| beta=False backward | 16 | 1.234x-1.377x | 1.278x | 16 | 0 |
| beta=False direct full | 16 | 1.094x-1.112x | 1.106x | 12 | 0 |
| beta=True backward | 16 | 0.792x-0.939x | 0.814x | 0 | 16 |
| beta=True direct full | 16 | 1.112x-1.134x | 1.121x | 16 | 0 |

Across the primary `M={1,16,128}` matrix, 12/24 backward pairs improved by at least 10% and 24/24 full pairs improved by at least 5%, but all 12 beta-present primary backward pairs regressed by more than 5%. This violates the hard gate regardless of the positive full results.

All 32 temporary reducer comparisons against the existing eager reductions passed at the gate tolerance (`rtol=5e-2`, `atol=1e-2`). This was a profile gate rather than final correctness qualification.

### Launch and traffic evidence

For BF16, beta=True, M=16, N=4096:

- The partial inputs contain 4,854,400 bytes: DG and DB are each `148 x 4096 x 4` bytes and DA is `148 x 8 x 4` bytes.
- PyTorch profiler: baseline targeted epilogue was three reductions plus three casts, totaling 20.061 us CUDA time; the single reducer was 38.849 us. The unchanged producer was 2.560 us baseline and 2.496 us candidate in this trace.
- NCU independently confirmed the targeted launch count changed 6 -> 1. Replay-instrumented targeted duration was 52.544 us for the six baseline launches versus 101.984 us for the reducer. The raw NCU files contain per-launch grid, duration, register, occupancy, and memory metrics.

The mechanism therefore exists, but the fixed fused launch performs too much serialized work when beta is present and is slower than the six highly optimized eager kernels it replaces.

### Reproduction and raw evidence

```bash
sbatch .cache/b300-dyt-reduction/hard_gate.sbatch
```

- Harness: `.cache/b300-dyt-reduction/hard_gate.py`
- Slurm wrapper: `.cache/b300-dyt-reduction/hard_gate.sbatch`
- Full rounds, summaries, profiler tables, and completion marker: `.cache/b300-dyt-reduction/gate-5034.out`
- Stderr: `.cache/b300-dyt-reduction/gate-5034.err` (only profiler start/stop notices)
- NCU baseline: `.cache/b300-dyt-reduction/gate-ncu-baseline-5034.csv`
- NCU candidate: `.cache/b300-dyt-reduction/gate-ncu-candidate-5034.csv`

The ignored harness passed Ruff, Ruff format check, Python byte-compilation, and `bash -n`. The repository worktree and index remained clean at the baseline commit. Full repository tests and `make checkstyle` were not run because the hard gate failed before any tracked implementation.

Job `5031` was canceled during its first round after detecting that its temporary full provider used a different autograd wrapper; all of its full-operation measurements are excluded. Job `5034` is the corrected, authoritative result.

Conclusion: **do not proceed with this fixed one-launch reducer and do not propose it upstream**. Per the issue contract, no alternate reducer schedule, producer tuning, fallback, or narrower matrix was attempted.
